man/man1/wc.1
.TH WC 1
.SH NAME
wc \- word count
.SH SYNOPSIS
.B wc
[
.B -lwrbc
]
[
.I file ...
]
.SH DESCRIPTION
.I Wc
counts lines, words, runes, syntactically-invalid
.SM UTF
codes and bytes in the named
.IR files ,
or in the standard input if no file is named.
A word is a maximal string of characters
delimited by spaces, tabs or newlines.
The count of runes includes invalid codes.
.PP
If the optional argument is present,
just the specified counts (lines, words, runes, broken
.SM UTF
codes or bytes)
are selected by the letters
.BR l ,
.BR w ,
.BR r ,
.BR b ,
or
.BR c .
Otherwise, lines, words and bytes
.RB ( -lwc )
are reported.
.SH SOURCE
.B /usr/local/plan9/src/cmd/wc.c
.SH BUGS
The Unicode Standard has many blank characters scattered through it,
but
.I wc
looks for only
.SM ASCII
space, tab and newline.
.br
.I Wc
should have options to count suboptimal
.SM UTF
codes
and bytes that cannot occur in any
.SM UTF
code.


CHANGES
November 26, 2005
	9term: add button 3 menu
	include: move utf.h, fmt.h from u.h to libc.h
	dist: print summary w/ errors during INSTALl

November 25, 2005
	root: add Makefile and configure for the blind,
		rename NOTES to README	

November 21, 2005
	lib9, 9pserve, vac: 9P2000.u fixes (Lucho Ionkov)

November 15, 2005
	libdraw: more X11 fixes
	libc.h: sched_yield NetBSD tweak (Anselm Garbe)

November 12, 2005
	faces: fix face location
	jpg, strings: remove references to /fd/

November 12, 2005
	various: X11 fixes

November 11, 2005
	9term: more fixes (Tim Wiess)

November 10, 2005
	9term: handle \r, other fixes (Tim Wiess)
	auxstats: new Darwin tweaks

November 9, 2005
	libthread: fix OS X context switch code

November 7, 2005
	vbackup: various fixes

November 6, 2005
	9term: new port directly from rio
	include/libc.h: p9syslog, not syslog in varargck
	libauthsrv: close fd, memory leak in authdial
	libdraw: X11 debugging helps
	lib9: add getcallerpc-arm (Ron Minnich)
	auxstats: Fix FreeBSD 6 build (William Josephson)

November 4, 2005
	stats: add -C to argchars (Lou Kamenov)
	libthread: ARM fixes (Ron Minnich), getcontext fixes

November 2, 2005
	vbackup: longer timeout on BSD mount
	libthread: Linux ARM fixes (Ron Minnich)

November 1, 2005
	various: ctype sign fixes (Tom Miller)
	libfmt: avoid overflow on %f
	libthread: Linux ARM fixes (Ron Minnich), FreeBSD 5 fixes
	libventi: various cleanup

October 31, 2005
	acme: preparation for multiline tags (Paul Lalonde)
	bc: 64-bit safe (Taj Khattra)
	faces: initial port (John Cummings)
	stats: add -F option (Lou Kamenov)
	various: ctype sign fixes (Tom Miller)

October 30, 2005
	auxstats: more OS X fixes (Jeff Sickel)	

October 29, 2005
	9p: clunk fids before exiting
	mk: handle MKSHELL type correctly
	upas, acme mail: add initial port (John Cummings)
	vbackup: silence compiler warnings
	venti: silence compiler warnings
	lib9: remove reference to ffork
		print correct file name in truerand error
	libdisk: add Disk routines
	libdiskfs: silence compiler warnings
	libframe: add noredraw flag
	libhtml: many fixes
	libsunrpc: silence compiler warnings

October 18, 2005
	9l: add autoframework (like autolib) for OS X (Jeff Sickel)
	page: use "9 file" always
	ps, psu: narrower display
	stats: better OS X support (Jeff Sickel)

October 17, 2005
	cb: undef old type macros before redefining.

October 5, 2005
	stats: better OS X support (Jeff Sickel) 

September 30, 2005
	libhtml: more &names and numbers (Erik Quanstrom)

September 29, 2005
	libc.h: add #pragmas for varargck, #ifdef'ed out.

September 26, 2005
	libthread: print thread info on SIGQUIT, SIGINFO, logging
	acme: add more thread names

September 21, 2005
	9term: fix SunOS typo (Bengt Kleberg, glaive)

September 19, 2005
	libhtml: more &names (Erik Quanstrom)

September 19, 2005
	lib9/truerand: OpenBSD lets you open /dev/random
		but not read it! (Chad Dougherty, Mechiel Lukkien)

September 15, 2005
	libdraw/x11-event: Ignore unexpected events (Rob Pike)
	jpg: do not revert to CMAP for second file (Rob Pike)

September 13, 2005
	9pserve: signedness bug

September 12, 2005
	dist: various attempts at Debian packaging
	lib/moveplan9.sh: allow specification of old directory
	Getdir: interpret tags with $ in them
	mpm: use g++ as LD, add tmac.pm to tmac directory
	include/libc.h, fcall.h: add 9P2000.u extensions
	9pserve: add 9P2000.u extensions
	vac: add 9P2000.u extensions, signedness bug

September 9, 2005
	lc: add -p like man page says
	various: insert (uchar) casts for ctype arguments
		for broken NetBSD.

September 4, 2005
	dist/main.html: note x86 OpenBSD

August 31, 2005
	file: spell IA-64 correctly (Zoltan Jarai)

August 30, 2005
	dist/debian, dist/pkg: start playing with building packages
	dist/isum: ignore 'loop not entered at top' (SunOS cc)
	dist/ss.html: add screen shots
	dist/unix.html: update VN license summary
	9l: no need to handle OpenBSD specially
	file: handle #! scripts, ELF better
	font/lucidasans/bold*unicode.7.font: more subfonts
	g: search .pl, .cgi
	hget: add hget(1)
	lp: add preliminary lp (not expected to work yet)
	src: use Plan 9 file(1)

August 12, 2005
	lib/moveplan9.files: add bin/stack
	include/libc.h: use DM constants from v9fs, add DMTMP
	libthread: add Linux-power-asm.s
	cb: add

August 11, 2005
	various: silence warnings
	unix/*: note new BSD license from Vita Nuova
	libdiskfs: add UFS2 support
	libthread: add NetBSD power assembly
	9term: send tty's current interrupt character on DEL keystroke
	rc: add limited rfork command, make sure errors propagate
	venti(8): document bloom line in config
	vftp: add get command
	vnfs: use strtoull for arguments instead of atoi.

August 10, 2005
	src/cmd/mkfile: do not build lp yet
	dist/addsrclinks, fixurls: use /usr/bin/env perl to find perl (NetBSD)
	lib9/_p9dir: avoid 32-bit overflow in old Linux size calculation
	lib9/truerand: use /dev/srandom on OpenBSD
	libthread: annoying wrappers for OpenBSD
	libmp: different assembly comments for old GNU as
	9p: set exit status
	file: add, appease gcc
	getflags: added
	mk: decide mtime-related race conservatively
	postscript/text2post: added
	rcmain: make fn cd safer for use with rc -e (in mk)
	rio: turn off debugging prints, remove 9wm.man, fixes from Andrey
	vbackup: NetBSD fixes for mount-BSD.c
	venti/srv: add <sys/stat.h> header for FreeBSD

July 28, 2005
	acme: subtle winunlock bug fix
	libdiskfs: fix double, triple indirect block reading for FFS
	vftp: new program
	vbackup: coalesce mount-*BSD files.

July 27, 2005
	include/u.h, src/libthread/thread.c: SunOS 10 attempt
	libdraw: less graceful, less raceful shutdown
	libthread: avoid C library races during threadexitsall
	awd: cannot use hostname -s on SunOS
	9pserve, srv: use correct type for NOFID

July 26, 2005
	various: fixes for SunOS from Bengt Kleberg

July 25, 2005
	vac: avoid infinite redirection while sniffing type of archive

July 24, 2005
	venti: many minor changes, updates

src/cmd/wc.c
/*
 * wc -- count things in utf-encoded text files
 * Bugs:
 *	The only white space characters recognized are ' ', '\t' and '\n', even though
 *	ISO 10646 has many more blanks scattered through it.
 *	Should count characters that cannot occur in any rune (hex f0-ff) separately.
 *	Should count non-canonical runes (e.g. hex c1,80 instead of hex 40).
 */
#include <u.h>
#include <libc.h>
#define	NBUF	(8*1024)
uvlong nline, tnline, pline;
uvlong nword, tnword, pword;
uvlong nrune, tnrune, prune;
uvlong nbadr, tnbadr, pbadr;
uvlong nchar, tnchar, pchar;
void count(int, char *);
void report(uvlong, uvlong, uvlong, uvlong, uvlong, char *);
void
main(int argc, char *argv[])
{
	char *status="";
	int i, f;
	ARGBEGIN {
	case 'l': pline++; break;
	case 'w': pword++; break;
	case 'r': prune++; break;
	case 'b': pbadr++; break;
	case 'c': pchar++; break;
	default:
		fprint(2, "Usage: %s [-lwrbc] [file ...]\n", argv0);
		exits("usage");
	} ARGEND
	if(pline+pword+prune+pbadr+pchar == 0) {
		pline = 1;
		pword = 1;
		pchar = 1;
	}
	if(argc==0)
		count(0, 0);
	else{
		for(i=0;i<argc;i++){
			f=open(argv[i], OREAD);
			if(f<0){
				perror(argv[i]);
				status="can't open";
			}
			else{
				count(f, argv[i]);
				tnline+=nline;
				tnword+=nword;
				tnrune+=nrune;
				tnbadr+=nbadr;
				tnchar+=nchar;
				close(f);
			}
		}
		if(argc>1)
			report(tnline, tnword, tnrune, tnbadr, tnchar, "total");
	}
	exits(status);
}
void
report(uvlong nline, uvlong nword, uvlong nrune, uvlong nbadr, uvlong nchar, char *fname)
{
	char line[1024], word[128];
	line[0] = '\0';
	if(pline){
		sprint(word, " %7llud", nline);
		strcat(line, word);
	}
	if(pword){
		sprint(word, " %7llud", nword);
		strcat(line, word);
	}
	if(prune){
		sprint(word, " %7llud", nrune);
		strcat(line, word);
	}
	if(pbadr){
		sprint(word, " %7llud", nbadr);
		strcat(line, word);
	}
	if(pchar){
		sprint(word, " %7llud", nchar);
		strcat(line, word);
	}
	if(fname){
		sprint(word, " %s",   fname);
		strcat(line, word);
	}
	print("%s\n", line+1);
}
/*
 * How it works.  Start in statesp.  Each time we read a character,
 * increment various counts, and do state transitions according to the
 * following table.  If we're not in statesp or statewd when done, the
 * file ends with a partial rune.
 *        |                character
 *  state |09,20| 0a  |00-7f|80-bf|c0-df|e0-ef|f0-ff
 * -------+-----+-----+-----+-----+-----+-----+-----
 * statesp|ASP  |ASPN |AWDW |AWDWX|AC2W |AC3W |AWDWX
 * statewd|ASP  |ASPN |AWD  |AWDX |AC2  |AC3  |AWDX
 * statec2|ASPX |ASPNX|AWDX |AWDR |AC2X |AC3X |AWDX
 * statec3|ASPX |ASPNX|AWDX |AC2R |AC2X |AC3X |AWDX
 */
enum{			/* actions */
	AC2,		/* enter statec2 */
	AC2R,		/* enter statec2, don't count a rune */
	AC2W,		/* enter statec2, count a word */
	AC2X,		/* enter statec2, count a bad rune */
	AC3,		/* enter statec3 */
	AC3W,		/* enter statec3, count a word */
	AC3X,		/* enter statec3, count a bad rune */
	ASP,		/* enter statesp */
	ASPN,		/* enter statesp, count a newline */
	ASPNX,		/* enter statesp, count a newline, count a bad rune */
	ASPX,		/* enter statesp, count a bad rune */
	AWD,		/* enter statewd */
	AWDR,		/* enter statewd, don't count a rune */
	AWDW,		/* enter statewd, count a word */
	AWDWX,		/* enter statewd, count a word, count a bad rune */
	AWDX		/* enter statewd, count a bad rune */
};
uchar statesp[256]={	/* looking for the start of a word */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 00-07 */
AWDW, ASP,  ASPN, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 08-0f */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 10-17 */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 18-1f */
ASP,  AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 20-27 */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 28-2f */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 30-37 */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 38-3f */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 40-47 */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 48-4f */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 50-57 */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 58-5f */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 60-67 */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 68-6f */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 70-77 */
AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW, AWDW,	/* 78-7f */
AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,/* 80-87 */
AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,/* 88-8f */
AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,/* 90-97 */
AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,/* 98-9f */
AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,/* a0-a7 */
AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,/* a8-af */
AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,/* b0-b7 */
AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,/* b8-bf */
AC2W, AC2W, AC2W, AC2W, AC2W, AC2W, AC2W, AC2W,	/* c0-c7 */
AC2W, AC2W, AC2W, AC2W, AC2W, AC2W, AC2W, AC2W,	/* c8-cf */
AC2W, AC2W, AC2W, AC2W, AC2W, AC2W, AC2W, AC2W,	/* d0-d7 */
AC2W, AC2W, AC2W, AC2W, AC2W, AC2W, AC2W, AC2W,	/* d8-df */
AC3W, AC3W, AC3W, AC3W, AC3W, AC3W, AC3W, AC3W,	/* e0-e7 */
AC3W, AC3W, AC3W, AC3W, AC3W, AC3W, AC3W, AC3W,	/* e8-ef */
AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,/* f0-f7 */
AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX,AWDWX/* f8-ff */
};
uchar statewd[256]={	/* looking for the next character in a word */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 00-07 */
AWD,  ASP,  ASPN, AWD,  AWD,  AWD,  AWD,  AWD,	/* 08-0f */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 10-17 */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 18-1f */
ASP,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 20-27 */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 28-2f */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 30-37 */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 38-3f */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 40-47 */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 48-4f */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 50-57 */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 58-5f */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 60-67 */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 68-6f */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 70-77 */
AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,  AWD,	/* 78-7f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 80-87 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 88-8f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 90-97 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 98-9f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* a0-a7 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* a8-af */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* b0-b7 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* b8-bf */
AC2,  AC2,  AC2,  AC2,  AC2,  AC2,  AC2,  AC2,	/* c0-c7 */
AC2,  AC2,  AC2,  AC2,  AC2,  AC2,  AC2,  AC2,	/* c8-cf */
AC2,  AC2,  AC2,  AC2,  AC2,  AC2,  AC2,  AC2,	/* d0-d7 */
AC2,  AC2,  AC2,  AC2,  AC2,  AC2,  AC2,  AC2,	/* d8-df */
AC3,  AC3,  AC3,  AC3,  AC3,  AC3,  AC3,  AC3,	/* e0-e7 */
AC3,  AC3,  AC3,  AC3,  AC3,  AC3,  AC3,  AC3,	/* e8-ef */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* f0-f7 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX	/* f8-ff */
};
uchar statec2[256]={	/* looking for 10xxxxxx to complete a rune */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 00-07 */
AWDX, ASPX, ASPNX,AWDX, AWDX, AWDX, AWDX, AWDX,	/* 08-0f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 10-17 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 18-1f */
ASPX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 20-27 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 28-2f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 30-37 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 38-3f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 40-47 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 48-4f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 50-57 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 58-5f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 60-67 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 68-6f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 70-77 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 78-7f */
AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR,	/* 80-87 */
AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR,	/* 88-8f */
AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR,	/* 90-97 */
AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR,	/* 98-9f */
AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR,	/* a0-a7 */
AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR,	/* a8-af */
AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR,	/* b0-b7 */
AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR, AWDR,	/* b8-bf */
AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X,	/* c0-c7 */
AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X,	/* c8-cf */
AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X,	/* d0-d7 */
AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X,	/* d8-df */
AC3X, AC3X, AC3X, AC3X, AC3X, AC3X, AC3X, AC3X,	/* e0-e7 */
AC3X, AC3X, AC3X, AC3X, AC3X, AC3X, AC3X, AC3X,	/* e8-ef */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* f0-f7 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX	/* f8-ff */
};
uchar statec3[256]={	/* looking for 10xxxxxx,10xxxxxx to complete a rune */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 00-07 */
AWDX, ASPX, ASPNX,AWDX, AWDX, AWDX, AWDX, AWDX,	/* 08-0f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 10-17 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 18-1f */
ASPX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 20-27 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 28-2f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 30-37 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 38-3f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 40-47 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 48-4f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 50-57 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 58-5f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 60-67 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 68-6f */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 70-77 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* 78-7f */
AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R,	/* 80-87 */
AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R,	/* 88-8f */
AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R,	/* 90-97 */
AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R,	/* 98-9f */
AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R,	/* a0-a7 */
AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R,	/* a8-af */
AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R,	/* b0-b7 */
AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R, AC2R,	/* b8-bf */
AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X,	/* c0-c7 */
AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X,	/* c8-cf */
AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X,	/* d0-d7 */
AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X, AC2X,	/* d8-df */
AC3X, AC3X, AC3X, AC3X, AC3X, AC3X, AC3X, AC3X,	/* e0-e7 */
AC3X, AC3X, AC3X, AC3X, AC3X, AC3X, AC3X, AC3X,	/* e8-ef */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX,	/* f0-f7 */
AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX, AWDX	/* f8-ff */
};
void
count(int f, char *name)
{
	int n;
	uchar buf[NBUF];
	uchar *bufp, *ebuf;
	uchar *state=statesp;

	nline = 0;
	nword = 0;
	nrune = 0;
	nbadr = 0;
	nchar = 0;

	for(;;){
		n=read(f, buf, NBUF);
		if(n<=0)
			break;
		nchar+=n;
		nrune+=n;	/* might be too large, gets decreased later */
		bufp=buf;
		ebuf=buf+n;
		do{
			switch(state[*bufp]){
			case AC2:   state=statec2;                   break;
			case AC2R:  state=statec2; --nrune;          break;
			case AC2W:  state=statec2; nword++;          break;
			case AC2X:  state=statec2;          nbadr++; break;
			case AC3:   state=statec3;                   break;
			case AC3W:  state=statec3; nword++;          break;
			case AC3X:  state=statec3;          nbadr++; break;
			case ASP:   state=statesp;                   break;
			case ASPN:  state=statesp; nline++;          break;
			case ASPNX: state=statesp; nline++; nbadr++; break;
			case ASPX:  state=statesp;          nbadr++; break;
			case AWD:   state=statewd;                   break;
			case AWDR:  state=statewd; --nrune;          break;
			case AWDW:  state=statewd; nword++;          break;
			case AWDWX: state=statewd; nword++; nbadr++; break;
			case AWDX:  state=statewd;          nbadr++; break;
			}
		}while(++bufp!=ebuf);
	}
	if(state!=statesp && state!=statewd)
		nbadr++;
	if(n<0)
		perror(name);
	report(nline, nword, nrune, nbadr, nchar, name);
}
